README.md
# Socket Input Params

A Node.js module used to receive input parameters via a Unix domain socket.

## Rationale

This module has been put together to work around a limitation of the [Meteor platform](https://meteor.com) that (as of this
 writing) does not provide a way for input parameters to be passed to the target (server) application.
 
It is also useful for passing sensitive data like passwords to running applications, and avoid them to be accidentally
 disclosed if other, more conventional mechanisms like command line arguments and environment variables are used.
 
## How It Works

The module opens a Unix domain socket named `inputparams` in the current work directory, and waits for data to be
 received through it. The received data is then parsed, and the resulting command line options are returned to the
 calling application.
 
## Installation

```shell
npm install socket-input-params
```

## Usage

### Instantiate the socket input parameters object.

```javascript
const SocketInputParams = require('socket-input-params');

const sockInParams = new SocketInputParams({
    readTimeout: 30000,
    optionDefs: [{
        name: 'password',
        alias: 'p'
    }]
});
```

#### Constructor options

The following options can be used when instantiating the socket input parameters object:

- **path** \[String\] - (optional, default: <b>*'.'*</b>) Path where the Unix domain socket should be created.
- **socketName** \[String\] - (optional, default: <b>*'inputparams'*</b>) Name of the Unix domain socket to be created.
- **readTimeout** \[Number\] - (optional, default: <b>*60000*</b>) Timeout, in milliseconds, for reading data from the
 Unix domain socket.
- **optionDefs** \[Array | Object\] - List of objects defining the command line options to be expected as input
 parameters as defined by the ['command-line-args'](https://github.com/75lb/command-line-args) Node.js module. When
 defining a single option, an object can be passed instead of an array.

### Retrieve command line options

#### Using a callback

```javascript
sockInParams.getCommandLineOptions((err, result) => {
    if (err) {
        console.error('Error retrieving command line options', err);
    }
    else {
        console.log('Retrieved command line options', result);
    }
})
```

#### Using promise

```javascript
sockInParams.getCommandLineOptions()
.then(result => {
    console.log('Retrieved command line options', result);
})
.catch(err => {
    console.error('Error retrieving command line options', err);
});
```

#### Sample returned command line options

```javascript
{
    password: '123456'
}
```

> **Note**: for more examples of how command line options are returned, please refer to the ['command-line-args'](https://github.com/75lb/command-line-args)
> Node.js module.

## Passing parameters to the running application

The following is a sample shell script that can be used to pass input parameters (via Unix domain socket) to a running
 application (that makes use of this module).

```shell script
#!/bin/bash
# Script used to send input parameters to running application via Unix domain socket
#
# How to use it:
#  . any parameters to be passed should be preceded by '--'
#  . if no password parameter is passed (-p <psw>) the script will ask for a password to be entered

cd "$APP_DIR"

# Filter internal (before '--') and external (after '--') parameters
idx=0
delimiterFound=0
hasPswOption=0

for arg; do
  if [ $delimiterFound -eq 1 ]; then
    # Save external parameter (to be passed to running application)
    extParams[idx]="$arg"

    if [[ $arg == "-p" || $arg == "--password" ]]; then
      hasPswOption=1
    fi
  else
    if [ $arg == "--" ]; then
      delimiterFound=1
    else
      # Save internal parameter (to be interpreted by this script)
      intParams[idx]="$arg"
    fi
  fi

  ((idx++))
done

if [ $hasPswOption -ne 1 ]; then
  # Request user to enter password (to be passed to running application)
  echo -n "Enter password: "
  read -s psw
  echo

  # Add password to external parameters
  extParams[idx]="-p"
  ((idx++))
  extParams[idx]="$psw"
fi

tmout=60

# Wait for Unix domain socket used to input parameters to be created (by running application)
while [ ! -S ./inputparams ]; do
  if [ $tmout -eq 0 ]; then
    echo "Timeout waiting for input parameters socket"
    exit -1
  fi
  sleep 1
  ((tmout--))
done

extParamsList=""

for param in "${extParams[@]}"; do
  extParamsList="$extParamsList '$param'"
done

# Send external parameters to running application (through open Unix domain socket)
echo -n "$extParamsList" | nc -U ./inputparams
```

> **Note**: it is assumed that the environment variable $APP_DIR contains the path of the directory where the Unix
> domain socket created by the running application is located.

## License

This Node.js module is released under the [MIT License](LICENSE). Feel free to fork, and modify!

Copyright © 2020, Blockchain of Things Inc.

CHANGELOG.md
# Changelog

## [1.0.2] - 2021-05-14

### Fixes
- Update dependency modules to mitigate security vulnerabilities.

## [1.0.1] - 2020-09-28

### Fixes
- Update information in Readme file and header comments.

## [1.0.0] - 2020-09-26

### New features
- Initial version.

test/testSocketInputParams.js
const net = require('net');
const fs = require('fs');
const expect = require('chai').expect;
const SocketInputParams = require('../index');

describe('Socket Input Parameters', function () {
    describe('Create new SocketInputParams instance', function () {
        it('should throw if no parameter is passed', function () {
            expect(() => {
                new SocketInputParams();
            }).to.throw(TypeError, 'SocketInputParams constructor: missing or invalid parameter \'opts\'');
        });

        it('should throw if a parameter of an inconsistent type is passed', function () {
            expect(() => {
                new SocketInputParams(0);
            }).to.throw(TypeError, 'SocketInputParams constructor: missing or invalid parameter \'opts\'');
        });

        it('should throw if options parameter does not contain required property', function () {
            expect(() => {
                new SocketInputParams({});
            }).to.throw(TypeError, 'SocketInputParams constructor: missing or invalid options property \'optionDefs\'');
        });

        it('should throw if property \'optionDefs\' of options parameter has an inconsistent type', function () {
            expect(() => {
                new SocketInputParams({optionDefs: 0});
            }).to.throw(TypeError, 'SocketInputParams constructor: missing or invalid options property \'optionDefs\'');
        });

        it('should succeed if only the required options are passed', function () {
            const sockInParams = new SocketInputParams({
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            expect(sockInParams.socketPath).to.equal('inputparams');
            expect(sockInParams.readTimeout).to.equal(60000);
        })

        it('should succeed when custom options are passed', function () {
            const sockInParams = new SocketInputParams({
                path: 'dir1/dir2',
                socketName: 'mysocket',
                readTimeout: 30000,
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            expect(sockInParams.socketPath).to.equal('dir1/dir2/mysocket');
            expect(sockInParams.readTimeout).to.equal(30000);
        })
    });

    describe('Receive input parameters (using callback)', function () {
        it('should throw if and invalid path is specified', function (done) {
            const sockInParams = new SocketInputParams({
                path: 'bla/bla/bla',
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            sockInParams.getCommandLineOptions((err, result) => {
                let error;

                try {
                    expect(result).to.not.exist;
                    expect(err).to.be.an.instanceOf(Error);
                    expect(err.message).to.match(/^Error reading data from Unix domain socket: /);
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });
        })

        it('should throw if received parameters do not match options definition', function (done) {
            const sockInParams = new SocketInputParams({
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            sockInParams.getCommandLineOptions((err, result) => {
                let error;

                try {
                    expect(result).to.not.exist;
                    expect(err).to.be.an.instanceOf(Error);
                    expect(err.message).to.match(/^Error parsing received input parameters: UNKNOWN_OPTION:/);
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });

            // Simulate passage of input parameters (with invalid option)
            net.createConnection(sockInParams.socketPath).end('-x');
        })

        it('should receive no input parameters if socket read times out', function (done) {
            const sockInParams = new SocketInputParams({
                readTimeout: 500,   // 500 ms
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            sockInParams.getCommandLineOptions((err, result) => {
                let error;

                try {
                    expect(err).to.not.exist;
                    expect(result).to.deep.equal({});
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });
        });

        it('should successfully retrieve command line options', function (done) {
            const sockInParams = new SocketInputParams({
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            sockInParams.getCommandLineOptions((err, result) => {
                let error;

                try {
                    expect(err).to.not.exist;
                    expect(result).to.deep.equal({
                        password: '123456'
                    });
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });

            // Simulate passage of input parameters
            net.createConnection(sockInParams.socketPath).end('-p 123456');
        });

        it('should succeed even if socket filename already exists', function (done) {
            const sockInParams = new SocketInputParams({
                readTimeout: 500,   // 500 ms
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            // Simulate existence of socket filename
            fs.writeFileSync(sockInParams.socketPath, '');

            sockInParams.getCommandLineOptions((err, result) => {
                let error;

                try {
                    expect(err).to.not.exist;
                    expect(result).to.deep.equal({
                        password: '123456'
                    });
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });

            setTimeout(() => {
                // Simulate passage of input parameters
                net.createConnection(sockInParams.socketPath).end('-p 123456');
            }, 100);
        });
    });

    describe('Receive input parameters (using promise)', function () {
        it('should throw if and invalid path is specified', function (done) {
            const sockInParams = new SocketInputParams({
                path: 'bla/bla/bla',
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            sockInParams.getCommandLineOptions()
            .then(result => {
                let error;

                try {
                    expect.fail('Function should not have succeeded');
                }
                catch (err) {
                    error = err;
                }

                done(error);
            })
            .catch(err => {
                let error;

                try {
                    expect(err).to.be.an.instanceOf(Error);
                    expect(err.message).to.match(/^Error reading data from Unix domain socket: /);
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });
        })

        it('should throw if received parameters do not match options definition', function (done) {
            const sockInParams = new SocketInputParams({
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            sockInParams.getCommandLineOptions()
            .then(result => {
                let error;

                try {
                    expect.fail('Function should not have succeeded');
                }
                catch (err) {
                    error = err;
                }

                done(error);
            })
            .catch(err => {
                let error;

                try {
                    expect(err).to.be.an.instanceOf(Error);
                    expect(err.message).to.match(/^Error parsing received input parameters: UNKNOWN_OPTION:/);
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });

            // Simulate passage of input parameters (with invalid option)
            net.createConnection(sockInParams.socketPath).end('-x');
        })

        it('should receive no input parameters if socket read times out', function (done) {
            const sockInParams = new SocketInputParams({
                readTimeout: 500,   // 500 ms
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });


            sockInParams.getCommandLineOptions()
            .then(result => {
                let error;

                try {
                    expect(result).to.deep.equal({});
                }
                catch (err) {
                    error = err;
                }

                done(error);
            })
            .catch(err => {
                let error;

                try {
                    expect.fail('Function should not have failed. Returned error: ' + err.toString());
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });
        });

        it('should successfully retrieve command line options', function (done) {
            const sockInParams = new SocketInputParams({
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            sockInParams.getCommandLineOptions()
            .then(result => {
                let error;

                try {
                    expect(result).to.deep.equal({
                        password: '123456'
                    });
                }
                catch (err) {
                    error = err;
                }

                done(error);
            })
            .catch(err => {
                let error;

                try {
                    expect.fail('Function should not have failed. Returned error: ' + err.toString());
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });

            // Simulate passage of input parameters
            net.createConnection(sockInParams.socketPath).end('-p 123456');
        });

        it('should succeed even if socket filename already exists', function (done) {
            const sockInParams = new SocketInputParams({
                readTimeout: 500,   // 500 ms
                optionDefs: {
                    name:'password',
                    alias: 'p'
                }
            });

            // Simulate existence of socket filename
            fs.writeFileSync(sockInParams.socketPath, '');

            sockInParams.getCommandLineOptions()
            .then(result => {
                let error;

                try {
                    expect(result).to.deep.equal({
                        password: '123456'
                    });
                }
                catch (err) {
                    error = err;
                }

                done(error);
            })
            .catch(err => {
                let error;

                try {
                    expect.fail('Function should not have failed. Returned error: ' + err.toString());
                }
                catch (err) {
                    error = err;
                }

                done(error);
            });

            setTimeout(() => {
                // Simulate passage of input parameters
                net.createConnection(sockInParams.socketPath).end('-p 123456');
            }, 100);
        });
    });
});

index.js
/**
 * Created by claudio on 2020-09-24
 */

const path = require('path');
const net = require('net');
const fs = require('fs');
const commandLineArgs = require('command-line-args');
const shellParser = require('shell-parser');

/**
 * Unix socket input parameters
 */
class SocketInputParams {
    /**
     * Class constructor
     * @param {Object} opts Options object with the following properties:
     * @param {String} [opts.path] Path where Unix domain socket used to receive parameters should be created.
     *                              If no path is passed, the current directory ('.') is used.
     * @param {String} [opts.socketName] Name of Unix domain socket to be created. If no name is passed, the
     *                                    name 'inputparams' is used.
     * @param {Number} [opts.readTimeout] Timeout, in milliseconds, for reading data from Unix domain socket.
     *                                     If no timeout is passed, a default 60 seconds timeout is used.
     * @param {Array|Object} opts.optionDefs List of objects defining the options to be expected as input parameters
     *                                        as defined by the 'command-line-args' Node.js module.
     */
    constructor(opts) {
        if (typeof opts !== 'object' || opts === null) {
            throw new TypeError('SocketInputParams constructor: missing or invalid parameter \'opts\'');
        }
        else if (typeof opts.optionDefs !== 'object' || opts.optionDefs === null) {
            throw new TypeError('SocketInputParams constructor: missing or invalid options property \'optionDefs\'');
        }

        this.socketPath = path.join(opts.path || '.', opts.socketName || 'inputparams');

        const readTimeout = Number.parseInt(opts.readTimeout);
        this.readTimeout = !Number.isNaN(readTimeout) ? readTimeout : 60000;

        this.optionsDefs = opts.optionDefs;
    }

    /**
     * Retrieve the command line options sent as input parameters over the Unix domain socket
     *
     * @param {Function} [callback] Callback function
     * @returns {Promise<Object>,undefined} If no callback is passed, a promise is returned. If the function succeeds,
     *                                       an object the properties of which are the parsed command line options
     *                                       is returned.
     */
    getCommandLineOptions(callback) {
        let result;

        // Prepare promise to be returned if no callback passed
        if (typeof callback !== 'function') {
            result = new Promise((resolve, reject) => {
                callback = (err, res) => {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(res);
                    }
                }
            });
        }

        // Prepare to receive input parameters
        this._receiveInputParams()
        .then(inputParams => {
            let cmdLineOpts;

            try {
                cmdLineOpts = this._parseInputParams(inputParams);
            }
            catch (err) {
                callback(err);
                return;
            }

            callback(null, cmdLineOpts);
        }, err => {
            callback(err);
        });

        return result;
    }

    async _receiveInputParams() {
        const dataLine = await this._readSocketData();

        if (fs.existsSync(this.socketPath)) {
            // Try to delete input parameters socket
            fs.unlink(this.socketPath, () => {});
        }

        return shellParser(dataLine);
    }

    _readSocketData() {
        return new Promise((resolve, reject) => {
            let readData = Buffer.from('');
            let connSocket;
            let toHandle;

            function returnSuccess() {
                if (toHandle) {
                    // Cancel timeout
                    clearTimeout(toHandle);
                }

                resolve(readData.toString());
            }

            function returnError(err) {
                if (toHandle) {
                    // Cancel timeout
                    clearTimeout(toHandle);
                }

                reject(err);
            }

            const server = net.createServer((socket) => {
                connSocket = socket;

                socket.on('data', (data) => {
                    readData = Buffer.concat([readData, data], readData.length + data.length);
                });

                socket.on('end', () => {
                    // Close server and return
                    server.close();
                    returnSuccess();
                });
            });

            let retry = false;

            server.once('error', (err) => {
                // Error reading from Unix domain socket. Check if socket already exists and
                //  try to delete it if so
                let socketDeleted = false;

                if (fs.existsSync(this.socketPath)) {
                    // Try to delete Unix domain socket
                    try {
                        fs.unlinkSync(this.socketPath);
                        socketDeleted = true;
                    }
                    catch(err) {
                    }
                }

                if (err.code === 'EADDRINUSE' && !retry && socketDeleted) {
                    // Try connecting to Unix domain socket again
                    retry = true;

                    server.close();
                    server.listen(this.socketPath);
                }
                else {
                    // Close connection and return error
                    server.close();
                    returnError(new Error('Error reading data from Unix domain socket: ' + err.toString()));
                }
            })

            server.listen(this.socketPath);

            toHandle = setTimeout(() => {
                toHandle = undefined;

                if (connSocket) {
                    connSocket.destroy();
                }

                // Close connection and return
                server.close();
                returnSuccess();
            }, this.readTimeout);
        });
    }

    _parseInputParams(inputParams) {
        // Save original process arguments list
        const procArgv = process.argv;
        const argv = [
            procArgv[0],
            procArgv[1]
        ];

        inputParams.forEach(param => {argv.push(param)});
        process.argv = argv;

        let cmdLineOpts;

        try {
            cmdLineOpts = commandLineArgs(this.optionsDefs);
        }
        catch (err) {
            throw new Error('Error parsing received input parameters: ' + err.toString());
        }

        // Restore original process arguments list
        process.argv = procArgv;

        return cmdLineOpts;
    }
}

module.exports = SocketInputParams;